src/rest.cpp

```diff
@@ -267,13 +262,57 @@ int Rest::Transpose(FunctorParams *functorParams)
     TransposeParams *params = dynamic_cast<TransposeParams *>(functorParams);
     assert(params);
 
-    // if (!this->HasOloc() || !this->HasPloc()) return FUNCTOR_SIBLINGS;
+    if ((!HasOloc() || !HasPloc()) && !HasLoc()) return FUNCTOR_SIBLINGS;
+
+    // Find whether current layer is top, middle (either one if multiple) or bttom
+    Staff *parentStaff = dynamic_cast<Staff *>(GetFirstAncestor(STAFF));
+    assert(parentStaff);
 
-    LogDebug("relocating rest");
+    Layer *parentLayer = dynamic_cast<Layer *>(GetFirstAncestor(LAYER));
+    assert(parentLayer);
 
-    TransPitch restLoc = this->GetTransLoc();
-    params->m_transposer->Transpose(restLoc);
-    this->UpdateFromTransLoc(restLoc);
+    Layer *firstLayer = dynamic_cast<Layer *>(parentStaff->GetFirst());
+    Layer *lastLayer = dynamic_cast<Layer *>(parentStaff->GetLast());
+
+    const bool isTopLayer = (firstLayer->GetN() == parentLayer->GetN());
+    const bool isBottomLayer = (lastLayer->GetN() == parentLayer->GetN());
+
+    // transpose based on @oloc and @ploc
+    if (HasOloc() && HasPloc()) {
+        const TransPitch centralLocation(6, 0, 4); // middle location of the staff
+        TransPitch restLoc(GetPloc() - PITCHNAME_c, 0, GetOloc());
+        params->m_transposer->Transpose(restLoc);
+        const bool isRestOnSpace = static_cast<bool>((restLoc.m_oct * 7 + restLoc.m_pname) % 2);
+        // on outer layers move rest on odd locations one line further 
+        if (isTopLayer && isRestOnSpace) {
+            restLoc++;
+        }
+        else if (isBottomLayer && isRestOnSpace) {
+            restLoc--;
+        }
+        if ((isTopLayer && (restLoc < centralLocation)) || (isBottomLayer && (restLoc > centralLocation))) {
+            restLoc = centralLocation;
+        }
+        
+        UpdateFromTransLoc(restLoc);
+    }
```

Comment:
I don't think this refinement is useful, but we can discuss more, and also with @lpugin.

The general problem is that if you want transposable rests, then you cannot quantize the ploc/oloc positions in the MEI data based on lines/spaces locations of the transposed rests, and the quantization has to be done only in the rendering process.   This is currently being done in verovio, but this code would break that behavior and probably limit the functionality of transposable rests? (I haven't compiled this code to test it).

Here is an example:

<img width="686" alt="Screen Shot 2020-08-11 at 8 57 25 AM" src="https://user-images.githubusercontent.com/3487289/89921407-8299ca80-dbb2-11ea-90b5-3271512021ee.png">

MEI encoding:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2020-08-11T08:57:43" version="3.0.0-dev-38985cb">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mdiv-0000000367288032">
    <score xml:id="score-0000000562863700">
     <scoreDef xml:id="scoredef-0000001022094311">
      <staffGrp xml:id="staffgrp-0000001509593138">
       <staffDef xml:id="staffdef-0000001581987438" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-0000000359584476" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <rest xml:id="rest-L4F1" dur="4" ploc="c" oloc="4" />
         <rest xml:id="rest-L5F1" dur="4" ploc="d" oloc="4" />
         <rest xml:id="rest-L6F1" dur="4" ploc="e" oloc="4" />
         <rest xml:id="rest-L7F1" dur="4" ploc="f" oloc="4" />
         <rest xml:id="rest-L8F1" dur="4" ploc="g" oloc="4" />
         <rest xml:id="rest-L9F1" dur="4" ploc="a" oloc="4" />
         <rest xml:id="rest-L10F1" dur="4" ploc="b" oloc="4" />
         <rest xml:id="rest-L11F1" dur="4" ploc="c" oloc="5" />
         <rest xml:id="rest-L12F1" dur="4" ploc="d" oloc="5" />
         <rest xml:id="rest-L13F1" dur="4" ploc="e" oloc="5" />
         <rest xml:id="rest-L14F1" dur="4" ploc="f" oloc="5" />
         <rest xml:id="rest-L15F1" dur="4" ploc="g" oloc="5" />
         <rest xml:id="rest-L16F1" dur="4" ploc="a" oloc="5" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

Notice that a quarter rest is placed on every diatonic step from C4 to A5 in the MEI data, but in the rendering, the spaces are avoided, so you see two rests in a row displayed at the same vertical position.  If such music were transposed up a step, the rests would not move (but this may depend on the top/bottom layer membership according to your code).  So when transposing the music up by a step 4 separate times compared to up a perfect fifth in one step, it seems that the ploc/oloc values of the rests would be different based on this code (while pitches which would not be different between the two different transposition processes).

Reply:
Well, this is certainly is strange, since this is the result that I get after rendering your MEI file:
![image](https://user-images.githubusercontent.com/1819669/89989469-ab40c380-dc89-11ea-8903-1dcfbfd3ced5.png)
Mind you, this is current behavior of the **develop** branch, i.e. without my changes and without transposition.
It's clearly different, so perhaps there's already some problems with rendering based on oloc/ploc?

Reply:
I think I did not re-render the MEI file that I was editing...